### ISSUE TITLE
Slices of pointers to structs seem to be mistranslated

### ISSUE THREAD

**Issue by RalfJung** [tchajed/goose#27]
A type like this
```
type KVClerkPool struct {
    mu *sync.Mutex
    // queue of free clerks
    freeClerks []*MemKVClerk
    coord HostName
}
```
is translated as
```
Definition KVClerkPool := struct.decl [
  "mu" :: lockRefT;
  "freeClerks" :: slice.T (struct.ptrT MemKVClerk);
  "coord" :: HostName
].
```
but indexing into that slice looks like
```
SliceGet (refT (struct.t MemKVClerk)) (struct.loadF KVClerkPool "freeClerks" "p") #0
```
Note that the slice element type is `refT (struct.t MemKVClerk)` here, but `struct.ptrT MemKVClerk` in the type declaration. I am not sure if those are the same type, but it doesn't seem like they are -- applying the `SliceGet` spec with an `is_slice` for element type `struct.ptrT MemKVClerk` fails.

For now I have worked around this by putting `*MemKVClerk` behind a `type` alias, but I don't even quite understand why that makes a difference.

**Comment by RalfJung** [comment]
> Note that the slice element type is refT (struct.t MemKVClerk) here, but struct.ptrT MemKVClerk in the type declaration. I am not sure if those are the same type, but it doesn't seem like they are 

They are not: `refT (struct.t MemKVClerk)` is `structRefT [struct.t MemKVClerk]`, i.e. a "pointer to a struct with a single field of type MemKVClerk", while `struct.ptrT MemKVClerk` is `structRefT (flatten_ty (structTy MemKVClerk))`, which actually recursively flattens the fields of `MemKVClerk`.

**Comment by RalfJung** [comment]
Interestingly, if `refT` was changed to
```coq
  Definition refT (t:ty) : ty := structRefT (flatten_ty t).
```
then the two things above would be the same. And that seems like a much more sensible definition anyway, from my understanding of how pointers work in GooseLang -- the pointee type should always be flattened, right? There is generally an implicit contract that in `structRefT tys`, no element of `tys` is a product or unit?

**Comment by RalfJung** [comment]
Turns out this is indeed fixed by using `flatten_ty` in `refT`: 2bd843435d4faeb7d56fe1318f9cd41dc8123a03, 46955dda18e2d7038a3cf41801262061a94e4215.

However, making the pointer type opaque is still a good idea, I think.
